resources/Microsoft.VSCode.DSC/Microsoft.VSCode.DSC.psm1

```diff
@@ -0,0 +1,115 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+$ErrorActionPreference = "Stop"
+Set-StrictMode -Version Latest
+
+enum VSCodeEnsure
+{
+    Absent
+    Present
+}
+
+#region DSCResources
+[DSCResource()]
+class VSCodeExtension
+{
+    [DscProperty(Key)]
+    [string] $Name
+
+    [DscProperty()]
+    [string] $Version
+
+    [DscProperty()]
+    [VSCodeEnsure] $Ensure = [VSCodeEnsure]::Present
+
+    [VSCodeExtension] Get()
+    {
+        $currentState = [VSCodeExtension]::new()
+        $currentState.Ensure = [VSCodeEnsure]::Absent
+        $currentState.Name = $this.Name
+        $currentState.Version = $this.Version
+
+        $installedExtensions = @{}
+        $extensionList = (Invoke-VSCode -Command "--list-extensions --show-versions") -Split [Environment]::NewLine
+
+        # Populate hash table with installed VSCode extensions.
+        foreach ($extension in $extensionList)
+        {
+            $info = $extension -Split '@'
+            $extensionName = $info[0]
+            $extensionVersion = $info[1]
+            $installedExtensions[$extensionName] = $extensionVersion
+        }
+
+        foreach ($extension in $installedExtensions.Keys)
+        {
+            if ($extension -eq $this.Name)
+            {
+                $currentState.Ensure = [VSCodeEnsure]::Present
+                $currentState.Name = $extension
+                $currentState.Version = $installedExtensions[$extension]
+                break
+            }
+        }
+        
+        return $currentState
+    }
+
+    [bool] Test()
+    {
+        $currentState = $this.Get()
+        if ($currentState.Ensure -ne $this.Ensure)
+        {
+            return $false
+        }
+
+        if ($null -ne $this.Version -and $this.Version -ne $currentState.Version)
+        {
+            return $false
+        }
+
+        return $true
+    }
+
+    [void] Set()
+    {
+        if ($this.Test())
+        {
+            return
+        }
+
+        if ($this.Ensure -eq [VSCodeEnsure]::Present)
+        {
+            $extensionArg = $this.Name
+
+            if ($null -ne $this.Version)
+            {
+                $extensionArg += "@$($this.Version)"
+            }
+
+            Invoke-VSCode -Command "--install-extension $($extensionArg)"
+        }
+        else
+        {
+            Invoke-VSCode -Command "--uninstall-extension $($this.Name)"
+        }
+    }
+}
+
+function Invoke-VSCode
+{
+    param (
+        [Parameter(Mandatory = $true)]
+        [string]$Command
+    )
+
+    try 
+    {
+        return Invoke-Expression "& `"$env:LocalAppData\Programs\Microsoft VS Code\bin\code.cmd`" $Command"
+    }
+    catch
+    {
+        throw "VSCode is not installed."
+    }
+}
```

Comment:
A few questions about this utility function:

- What if `code` is installed, but not in this location?
- What happens if the command is given invalid flags?
- What happens if the command is given valid flags but fails for another reason (e.g. can't connect to the internet)?
- What output does this return? Can that output be meaningfully parsed?
- Is this utility function intended for use for other resources managing VS Code, or just this one? Can this function shift any of the logic for defining arguments into the function?
- If you find it easier to test functions than classes in Pester, you could extract the logic for getting all extensions, installing an extension, and uninstalling an extension into functions that accept the name/version parameters and pass the instance properties or pipe the instance to the function, i.e.:

  ```powershell
  function Install-VSCodeExtension
  {
      [CmdletBinding()]
      param (
        [Parameter(Mandatory, ValueFromPipelineByPropertyName)]
        [string]$Name,
        [Parameter(ValueFromPipelineByPropertyName)]
        [string]$Version
      )
      
      # Implementation
  }
  
  $extension = [VSCodeExtension]@{
      Name    = 'Foo'
      Version = '1.2.3'
  }
  
  # Direct pass:
  Install-VSCodeExtension -Name $extension.Name -Version $extension.Version
  
  # Pipeline:
  $extension | Install-VSCodeExtension
  ```

Reply:
Code can only be installed in two locations based on scope. 
`C:\Users\<Your Username>\AppData\Local\Programs\Microsoft VS Code`
`C:\Program Files\Microsoft VS Code`

Code.exe doesn't return a proper error message if you provide invalid arguments or even an invalid extension name so it is difficult to produce a reasonable error message from the output. I agree that this is another concern that I have. 

The utility functions that we have here are intended to be used for other DSC resources for VSCode. This prototype that I have here is just for adding VSCode extensions, but the code.exe cli tool is very powerful so I'm sure there are more things we can add to this DSC module.

Changed to suggested functions for easier testing.

Reply:
Those are the two locations for VS Code _on Windows when installed with WinGet_, right? But the logic for installing/updating VS Code extensions is the same when a user installs VS Code in [portable mode] or on macOS/Linux - I'm not too worried about this, but it's worth keeping in mind that aside from this code snippet, everything else I see in here works fully across platforms and install methodologies, and PSDSC resources published to the gallery, especially for managing something like VS Code, are likely to draw users this code won't support.

I think both deferring this problem and/or just saying "Windows only" support for the module and resources are okay, to be clear.

[portable mode]: https://code.visualstudio.com/docs/editor/portable

Reply:
Good callout. I left a todo comment on the function stating that this is only supported on user/machine installs for 'Windows'. I'll need to figure out a better way to determine the install location especially to support the portable install scenario.